src/server/pkg/work/work.go

```diff
@@ -83,140 +100,196 @@ func (m *Master) Run(ctx context.Context, task *Task) (retErr error) {
 // The processFunc callback will be called for each subtask that needs to be processed
 // in the task.
 type Worker struct {
-	etcdClient          *etcd.Client
-	taskCol, subtaskCol col.Collection
-	processFunc         ProcessFunc
+	*taskEtcd
+	taskQueue *taskQueue
 }
 
 // NewWorker creates a new worker.
-func NewWorker(etcdClient *etcd.Client, etcdPrefix string, processFunc ProcessFunc) *Worker {
+func NewWorker(ctx context.Context, etcdClient *etcd.Client, etcdPrefix string, taskNamespace string) *Worker {
 	return &Worker{
-		etcdClient:  etcdClient,
-		taskCol:     collection(etcdClient, path.Join(etcdPrefix, taskPrefix)),
-		subtaskCol:  collection(etcdClient, path.Join(etcdPrefix, subtaskPrefix)),
-		processFunc: processFunc,
+		taskEtcd:  newTaskEtcd(etcdClient, etcdPrefix, taskNamespace),
+		taskQueue: newTaskQueue(ctx),
 	}
 }
 
+// ProcessFunc is a callback that is used for processing a subtask in a task.
+type ProcessFunc func(context.Context, *Task, *Task) error
+
 // Run runs the worker with the given context.
 // The worker will continue to watch the task collection until the context is cancelled.
-func (w *Worker) Run(ctx context.Context) error {
+func (w *Worker) Run(ctx context.Context, processFunc ProcessFunc) error {
 	return w.taskCol.ReadOnly(ctx).WatchF(func(e *watch.Event) (retErr error) {
-		var id string
-		taskInfo := &TaskInfo{}
-		if err := e.Unmarshal(&id, taskInfo); err != nil {
+		var taskID string
+		task := &Task{}
+		if err := e.Unmarshal(&taskID, task); err != nil {
 			return err
 		}
-		if taskInfo.State == State_RUNNING {
-			return w.processTask(ctx, taskInfo.Task)
+		if e.Type == watch.EventDelete {
+			return w.taskQueue.deleteTask(taskID)
+		}
+		taskChans, err := w.taskQueue.createTask(ctx, taskID)
+		if err != nil {
+			return err
 		}
+		go func() {
+			if err := w.taskFunc(task, taskChans, processFunc); err != nil && taskChans.ctx.Err() != context.Canceled {
+				// (bryce) how should logging work?
+				fmt.Printf("errored in task callback: %v\n", err)
+			}
+		}()
 		return nil
 	})
 }
 
-func (w *Worker) processTask(ctx context.Context, task *Task) error {
-	var eg errgroup.Group
-	ctx, cancel := context.WithCancel(ctx)
-	// Watch for task termination.
-	// This will handle completion, failure, and cancellation.
-	eg.Go(func() error {
-		return w.taskCol.ReadOnly(ctx).WatchOneF(task.Id, func(e *watch.Event) error {
-			var id string
-			taskInfo := &TaskInfo{}
-			if err := e.Unmarshal(&id, taskInfo); err != nil {
+func (w *Worker) taskFunc(task *Task, taskChans *taskChans, processFunc ProcessFunc) error {
+	claimWatch, err := w.claimCol.ReadOnly(taskChans.ctx).WatchOne(task.ID, watch.WithFilterPut())
+	if err != nil {
+		return err
+	}
+	subtaskWatch, err := w.subtaskCol.ReadOnly(taskChans.ctx).WatchOne(task.ID, watch.WithFilterDelete())
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case e := <-claimWatch.Watch():
+			if e.Type == watch.EventError {
 				return err
```

Comment:
I think this (and line 235) should be `return e.Err`

Reply:
Fixed.